lua/bufferline/events.lua

```diff
@@ -0,0 +1,248 @@
+
+  create_autocmd('BufModifiedSet', {
+    callback = function(tbl)
+      local is_modified = buf_get_option(tbl.buf, 'modified')
+      if is_modified ~= vim.b[tbl.buf].checked then
+        buf_set_var(tbl.buf, 'checked', is_modified)
+        render.update()
+      end
+    end,
+    group = augroup_render,
+  })
+
+  create_autocmd({'BufEnter', 'BufNew'}, {
+    callback = function() render.update(true) end,
+    group = augroup_render,
+  })
+
+  create_autocmd(
+    {
+      'BufEnter', 'BufWinEnter', 'BufWinLeave', 'BufWritePost',
+      'DiagnosticChanged',
+      'TabEnter',
+      'VimResized',
+      'WinEnter', 'WinLeave',
+    },
+    {
+      callback = function() render.update() end,
+      group = augroup_render,
+    }
+  )
+
+  create_autocmd('OptionSet', {
+    callback = function() render.update() end,
+    group = augroup_render,
+    pattern = 'buflisted',
+  })
+
+  create_autocmd('SessionLoadPost', {
+    -- TODO: I'm not sure if this whole thing can just be thrown in a `schedule_wrap`
+    --       in order to remove the inner `defer_fn` and `schedule` calls
+    callback = function()
+      if state.loading_session then
+        return
+      end
+
+      local restore_cmd = vim.g.Bufferline__session_restore
+      if restore_cmd then command(restore_cmd) end
+
+      vim.defer_fn(function() state.loading_session = false end, 100)
+      schedule(function() render.update(true) end)
+    end,
+    group = augroup_render,
+  })
+
+  create_autocmd('TermOpen', {
+    callback = function() defer_fn(function() render.update(true) end, 500) end,
+    group = augroup_render,
+  })
+
+  create_autocmd('User', {
+    callback = function()
+      -- We're allowed to use relative paths for buffers iff there are no tabpages
+      -- or windows with a local directory (:tcd and :lcd)
+      local use_relative_file_paths = true
+
+      -- PERF: I didn't import `haslocaldir` since it is only used here (just before calling `:mksession` and exiting vim)
+      local haslocaldir = vim.fn.haslocaldir --- @type function
+
+      for tabnr, tabpage in ipairs(list_tabpages()) do
+        if not use_relative_file_paths or haslocaldir(-1, tabnr) == 1 then
+          use_relative_file_paths = false
+          break
+        end
+
+        for _, win in ipairs(tabpage_list_wins(tabpage)) do
+          if haslocaldir(win, tabnr) == 1 then
+            use_relative_file_paths = false
+            break
+          end
+        end
+      end
+
+      local buffers = {}
+      for _, bufnr in ipairs(state.buffers) do
+        local name = buf_get_name(bufnr)
+        if use_relative_file_paths then
+          name = utils.relative(name)
+        end
+
+        -- escape quotes
+        table_insert(buffers, {name = name, pinned =  state.get_buffer_data(bufnr).pinned})
+      end
+
+      vim.g.Bufferline__session_restore = "lua require'bufferline.state'.restore_buffers " ..
+        vim.inspect(buffers, {newline = ' ', indent = ''})
```

Comment:
We were trying to convert a Lua dictionary to a string by ourselves, but `vim.inspect` does that for us (and handles all the edge cases of obscurely-named files).